Add an off-by-default warning to complain about MSVC bitfield padding

ojhunt opened the pull request:
This just adds a warning for bitfields placed next to other bitfields where the underlying type has different storage. Under the MS struct bitfield packing ABI such bitfields are not packed.

ojhunt committed:
Add an off-by-default warning to complain about MSVC bitfield padding

ojhunt commented:
Oh, my goal for this warning is to have us turn it on in the llvm+clang builds, so we can then move away from endless use of unsigned bitfields and just use enums as nature intended :D (cc @AaronBallman )

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6418,6 +6418,12 @@ def warn_signed_bitfield_enum_conversion : Warning<
+def warn_ms_bitfield_mismatched_storage_packing : Warning<
+  "bit-field %0 of type %1 has a different storage size (%2 vs %3 bytes) than the "
+  "preceding bit-field and may not be packed under MSVC ABI">,
+  InGroup<MSBitfieldCompatibility>, DefaultIgnore;
```
This seems like a warning we should probably enable by default when the user specifies `-fms-compatibility` or is compiling for an MSVC target. WDYT @rnk?

Sirraide replied:
I was about to point this out too but you beat me to it ;Þ

rnk replied:
If we did do this, it would be a Windows-thing, not an MSVC-thing, since it affects mingw. cc @mstorsjo  for a mingw PoV

I don't think we should enable it by default until we have more confidence that there aren't false positives. I don't see that as a blocker to land it, though.

mstorsjo replied:
Yes; mingw also uses the MS bitfield packing logic. (Originally, for mingw targets, users would need to opt in to this with `-mms-bitfields`; GCC enabled this option by default for mingw targets in 4.7 and we followed suit a bit belatedly in Clang 11 - see e3fd9dc9734c5775dc6824d0a839702e8d43e7f6.)

As for whether this should be enabled by deault; I think it's entirely up to each individual project whether they want it as an early portability headsup that they're writing code which is behaving differently on Windows or not. People writing windows-exclusive code might not care, and they may want to intentionally keep their structs as they are to match some existing ABI.

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19213,6 +19213,29 @@ void Sema::ActOnFields(Scope *S, SourceLocation RecLoc, Decl *EnclosingDecl,
+    auto IsNonDependentBitField = [](const FieldDecl *FD) {
+      if (!FD->isBitField())
+        return false;
+      if (FD->getType()->isDependentType())
+        return false;
+      return true;
```
```suggestion
      return !FD->getType()->isDependentType();
```

Sirraide replied:
Honestly, I’d even do this:
```c++
auto IsNonDependentBitField = [](const FieldDecl *FD) {
  return FD->isBitField() && !FD->getType()->isDependentType();
};
```
(I would have loved to format that as suggestion, but github isn’t letting me because this is in a reply...)

ojhunt replied:
I think I did this intentionally as I sometimes feel like explicit true/false returns are clearer, I'm happy with either Aaron or @Sirraide's suggestion but I want to confirm that others don't have my clarity concerns :D

Sirraide commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6418,6 +6418,12 @@ def warn_signed_bitfield_enum_conversion : Warning<
 def note_change_bitfield_sign : Note<
   "consider making the bitfield type %select{unsigned|signed}0">;
+def warn_ms_bitfield_mismatched_storage_packing : Warning<
+  "bit-field %0 of type %1 has a different storage size (%2 vs %3 bytes) than the "
+  "preceding bit-field and may not be packed under MSVC ABI">,
```
```suggestion
  "bit-field %0 of type %1 has a different storage size than the "
  "preceding bit-field (%2 vs %3 bytes) and will not be packed under "
  "the MSVC ABI">,
```
I think it’s a bit clearer this way.

ojhunt replied:
Thanks, I'm terrible at wording (I prodded discord specifically to get diagnostic wording)

rnk commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19213,6 +19213,29 @@ void Sema::ActOnFields(Scope *S, SourceLocation RecLoc, Decl *EnclosingDecl,
     if (Record && FD->getType().isVolatileQualified())
       Record->setHasVolatileMember(true);
+    auto IsNonDependentBitField = [](const FieldDecl *FD) {
+      if (!FD->isBitField())
```
Is this better as `return FD->isBitField() && !FD->getType()->isDependentType();`?

AaronBallman reviewed (commented):
Thank you for working on this! Please be sure to also add a release note to `clang/docs/ReleaseNotes.rst` so users know about the new diagnostic.

rnk commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19001,9 +19001,9 @@ void Sema::ActOnFields(Scope *S, SourceLocation RecLoc, Decl *EnclosingDecl,
   // Verify that all the fields are okay.
   SmallVector<FieldDecl*, 32> RecFields;
-
+  std::optional<const FieldDecl *> PreviousField;
```
It doesn't seem idiomatic to use an optional pointer, when nullable pointers are built into the language. Consider helping to reduce our collective carbon footprint by saving an extra template instantiation today, and using nullptr instead. =D

rnk commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19213,6 +19213,29 @@ void Sema::ActOnFields(Scope *S, SourceLocation RecLoc, Decl *EnclosingDecl,
+    if (Record && PreviousField && IsNonDependentBitField(FD) &&
+        IsNonDependentBitField(*PreviousField)) {
+      unsigned FDStorageSize =
+          Context.getTypeSizeInChars(FD->getType()).getQuantity();
```
`CharUnits` are 64-bit, so you've got an implicit truncation here, not that anyone would ever define an `i(1<<32)` type and use it in a bit field. However, I think it's cleaner and more straightforward to skip `getQuantity` and just compare the CharUnits directly for equality. I'm also curious if the diagnostic printer can handle CharUnits, maybe it can, but if not, that's only one extra getQuantity call.

rnk commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -631,6 +631,7 @@ def Packed : DiagGroup<"packed", [PackedNonPod]>;
 def PaddedBitField : DiagGroup<"padded-bitfield">;
 def Padded : DiagGroup<"padded", [PaddedBitField]>;
 def UnalignedAccess : DiagGroup<"unaligned-access">;
+def MSBitfieldCompatibility : DiagGroup<"ms-bitfield-packing-compatibility">;
```
Bike shedding: Should we drop "packing"? Note that this covers compatibility with GCC `-mms-bitfields` and structs carrying `__attribute__((ms_struct))`, so naming parallel to `-mms-bitfields` seems good.

I think you could make this warning much more discoverable if you add documentation for it in DiagnosticDocs.td, which will be rendered to https://clang.llvm.org/docs/DiagnosticsReference.html , which is highly searchable.

ADKaster commented:
https://github.com/llvm/llvm-project/pull/70978, which split -Wpadded into -Wpadded and -Wpadded-bitfield, may be relevant here

DanShaders commented:
Just FYI, there's a similar diagnostic `-Wpadded-bitfield` that warns about any padding inserted into bitfields in Itanium mode. They are somehow emitted in the completely different parts of the layout code, so it doesn't show up in the diff at all.

ojhunt committed:
Merge remote-tracking branch 'origin/llvm.org/main' into users/ojhunt/warn-msvc-bitfield-packing

ojhunt committed:
Address feedback

ojhunt committed:
Documentation and renaming

ojhunt committed:
Add release note

ojhunt committed:
Merge branch 'main' into users/ojhunt/warn-msvc-bitfield-packing

Sirraide commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -633,6 +633,50 @@ def Packed : DiagGroup<"packed", [PackedNonPod]>;
+    where you cannot specify the underlying type, the options are to either switch
+    to int sized storage for all specifiers or to resort to declaring the
+    bit-fields with explicit integer storage types and cast in and out of the field.
+    If such a solution is required the `preferred_type` attribute can be used to
+    convey the actual field type to debuggers and other tooling.
```
Can we link to the documentation of that attribute here maybe?

Sirraide reviewed (commented):
So I take it we decided not to enable it by default in `-fms-compatibility` mode then?

ojhunt committed:
Add links to other docs

ojhunt commented:
> So I take it we decided not to enable it by default in `-fms-compatibility` mode then?

I don't believe it is appropriate to do so. The intent of this warning is to indicate MSVC compatibility issues when building in non-ms-compatibility modes. The existing padding warnings already trigger for these layouts in the relevant ms compatibility modes

rnk commented on clang/test/SemaCXX/ms_struct-bitfield-padding.cpp:
```diff
@@ -0,0 +1,180 @@
+};
+
+static_assert(sizeof(M) == 4);
+
```
Can we add a test for a case for what happens when there is no layout difference, so something like this:
```
struct Foo {
  char a : 4;
  char b : 4;
  char c : 4;
  char d : 4;
  short x : 7;
  short y : 9;
};
```

As written, this code will warn, but both compilers use the same / similar struct layouts ([link](https://godbolt.org/z/qf6ccq65f)). Does this represent a false positive, or should we encourage users to use a consistent bitfield width?

`llvm::Value` doesn't have this problem because it happens to alternate between small bitfields and small integers, but this seems like a category of false positive that is hard to identify.

The only reasonable way to overcome this limitation would be to move this warning into record layout, in which case we'd have to implement it twice, once in the MS and non-MS record layout code.

ojhunt replied:
@rnk thoughts?

rnk replied:
Wouldn't this negatively affect cases like Value, where we have alternating small integeters and bitfields? The current layout won't be affected, but it seems like you'll get warnings on cases like this that you can't fix without adding padding:

```
struct Value {
  char kind;
  unsigned char b1 : 1;
  unsigned char b7 : 7; // 8
  unsigned short b2 : 2;
  unsigned short b14 : 14; // 16
  unsigned int b12 : 12;
  unsigned int b10a : 10;
  unsigned int b10b : 10; // 32
};
```

Does your warning fire on this case, and what types would one use to silence the warning and get compact layouts on all platforms?

rnk commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -631,6 +631,9 @@ Improvements to Clang's diagnostics
 
 - Clang now diagnoses dangling references for C++20's parenthesized aggregate initialization (#101957).
 
+- A new off-by-default warning ``-Wms-bitfield-compatibility`` has been added to alert to cases where bit-field
```
I wanted to bikeshed the name. Is this really a compatibility warning? It's a compatibility issue for someone trying to use C structs to do memory-mapped I/O or something, but for LLVM's application, it's supposed to be a warning about extra bitfield padding, suggesting a name like `-Wms-bitfield-padding` or `-Wms-bitfield-packing`.

It also doesn't catch all bitfield layout differences. There are cases like [this](https://godbolt.org/z/hvd8aY8ba). Do we care about that, or is it better to leave that to `-Wpadding`?
```
struct Foo {
  char a;
  int x : 7; // msvc aligns to 4, gcc doesn't
  int y : 9;
};
```

ojhunt committed:
Address feedback

AaronBallman commented:
> > So I take it we decided not to enable it by default in `-fms-compatibility` mode then?
> 
> I don't believe it is appropriate to do so. The intent of this warning is to indicate MSVC compatibility issues when building in non-ms-compatibility modes. The existing padding warnings already trigger for these layouts in the relevant ms compatibility modes

We don't typically add new off-by-default warnings though; users don't enable them often enough to be worth the costs. My thought process is: if we enable this diagnostic by default in `-fms-compatibility` mode as telling the user their bit-fields aren't packing, then it's on by default for some configurations so it meets our bar for inclusion, and it helps us directly because we have Windows precommit (and post-commit) CI coverage for building Clang and LLVM. Users on other platforms can opt-in to the diagnostic if they want the diagnostics for compatibility reasons.

If we want to leave it off by default, then I wonder if we want to roll the functionality into `-Wpadded-bitfield` which already exists and is off by default.

ojhunt committed:
Merge branch 'main' into users/ojhunt/warn-msvc-bitfield-packing

ojhunt committed:
Update documentation label

ojhunt commented:
> > > So I take it we decided not to enable it by default in `-fms-compatibility` mode then?
> > 
> > 
> > I don't believe it is appropriate to do so. The intent of this warning is to indicate MSVC compatibility issues when building in non-ms-compatibility modes. The existing padding warnings already trigger for these layouts in the relevant ms compatibility modes
> 
> We don't typically add new off-by-default warnings though; users don't enable them often enough to be worth the costs. My thought process is: if we enable this diagnostic by default in `-fms-compatibility` mode as telling the user their bit-fields aren't packing, then it's on by default for some configurations so it meets our bar for inclusion, and it helps us directly because we have Windows precommit (and post-commit) CI coverage for building Clang and LLVM. Users on other platforms can opt-in to the diagnostic if they want the diagnostics for compatibility reasons.

My concerns with attaching it to `-fms-compatibility` is that projects using that and that care about padding already have the padding/packing warnings enabled. For every other user we will be warning them that the padding matches the abi they're targeting. The problem I'm wanting to address is not "I am targeting the ms abi" it is "I am not targeting ms abi but want to be told about packing/padding that would be different on ms platforms".

The intention would be to then enable the warning (maybe with a `-Werror=...` once the build is clean :D ) for all llvm projects, not just the windows targets - maybe with some evangelism so larger projects that have similar issues can be made aware of the flag. Then we can start migrating away from unending unsigned bit-fields and adopt enum bit-fields

> 
> If we want to leave it off by default, then I wonder if we want to roll the functionality into `-Wpadded-bitfield` which already exists and is off by default.

The problem with this is that that people who do not care about windows/ms abi will then be getting what to them are spurious warnings.

ojhunt committed:
Merge remote-tracking branch 'origin/llvm.org/main' into users/ojhunt/warn-msvc-bitfield-packing

ojhunt committed:
Merge branch 'main' into users/ojhunt/warn-msvc-bitfield-packing

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -19240,6 +19240,25 @@ void Sema::ActOnFields(Scope *S, SourceLocation RecLoc, Decl *EnclosingDecl,
+      return FD->isBitField() && !FD->getType()->isDependentType();
+    };
+
+    if (Record && PreviousField && IsNonDependentBitField(FD) &&
```
If this is not enabled by default, maybe we should check the diagnostic is not ignored to avoid unnecessary work (`DiagnosticsEngine::isIgnored`)

ojhunt committed:
Merge branch 'main' into users/ojhunt/warn-msvc-bitfield-packing

AaronBallman commented:
> > > > So I take it we decided not to enable it by default in `-fms-compatibility` mode then?
> > > 
> > > 
> > > I don't believe it is appropriate to do so. The intent of this warning is to indicate MSVC compatibility issues when building in non-ms-compatibility modes. The existing padding warnings already trigger for these layouts in the relevant ms compatibility modes
> > 
> > 
> > We don't typically add new off-by-default warnings though; users don't enable them often enough to be worth the costs. My thought process is: if we enable this diagnostic by default in `-fms-compatibility` mode as telling the user their bit-fields aren't packing, then it's on by default for some configurations so it meets our bar for inclusion, and it helps us directly because we have Windows precommit (and post-commit) CI coverage for building Clang and LLVM. Users on other platforms can opt-in to the diagnostic if they want the diagnostics for compatibility reasons.
> 
> My concerns with attaching it to `-fms-compatibility` is that projects using that and that care about padding already have the padding/packing warnings enabled. For every other user we will be warning them that the padding matches the abi they're targeting. The problem I'm wanting to address is not "I am targeting the ms abi" it is "I am not targeting ms abi but want to be told about packing/padding that would be different on ms platforms".
> 
> The intention would be to then enable the warning (maybe with a `-Werror=...` once the build is clean :D ) for all llvm projects, not just the windows targets - maybe with some evangelism so larger projects that have similar issues can be made aware of the flag. Then we can start migrating away from unending unsigned bit-fields and adopt enum bit-fields
> 
> > If we want to leave it off by default, then I wonder if we want to roll the functionality into `-Wpadded-bitfield` which already exists and is off by default.
> 
> The problem with this is that that people who do not care about windows/ms abi will then be getting what to them are spurious warnings.

My opinion on this is evolving -- this morally is similar to all the padding warnings we have in that it's telling you "hey, this packing is suboptimal". Those warnings are all off-by-default because they're not identifying a logical mistake in your code; your code still *works* even though it may be possible to improve it. So I think leaving this as an off-by-default warning is reasonable.

And it is a distinct diagnostic from `-Wpadded-bitfield` in that it's identifying new problematic patterns which may not be problematic on all platforms, so it should get its own warning group. Maybe that warning group sits under `-Wpadded-bitfield`, but we should give users a way to say "I want to know about bitfield padding issues, but not ones specific to how Visual Studio does things".

ojhunt committed:
Merge branch 'main' into users/ojhunt/warn-msvc-bitfield-packing

ojhunt committed:
Don't try to diagnose the mismatch unless the warning is enabled

No longer perform the field analysis unless the warning is enabled.

For safety i've updated a few of the existing bitfield tests to also run
with this diagnostic enabled just to get a bit more coverage given we're
no longer running this diagnostic code for every test.

ojhunt committed:
Sigh i really thought i had the formatting correct

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6519,6 +6519,13 @@ def warn_signed_bitfield_enum_conversion : Warning<
+def warn_ms_bitfield_mismatched_storage_packing : Warning<
+  "bit-field %0 of type %1 has a different storage size than the "
+  "preceding bit-field (%2 vs %3 bytes) and will not be packed under "
+  "the MS Windows platform ABI">,
```
```suggestion
  "the Microsoft ABI">,
```

rnk replied:
In the codebase, we tend to use "Windows" to cover behaviors that need to be shared with GCC and MinGW, and "Microsoft" to cover behaviors that are unique to MSVC. This tends to boil down to C vs. C++. If it's part of the C ABI, it's a Windows behavior, under the rationale that any C compiler for the platform would need to match this behavior, like [enums always using a fixed type of `int`](https://godbolt.org/z/P41YYGEah) rather than expanding to `long long` if the members require it. C++ behaviors are high cost and "optional".

This mostly reflects the actual coding logic required, which is to check whether the OS is Windows, or whether the "environment" is Microsoft, so perhaps this is not the best language to communicate with users

AaronBallman replied:
Thanks for the explanation!

Would you recommend we go with `Windows ABI` and leave the `Microsoft` off? Or should we use `Microsoft Windows ABI`? I mostly want to avoid `MS`.

rnk replied:
In this context I would just say "Windows".

I do recall preferring the use of the company name ("Microsoft") over the use of the product name ("MSVC" / "Visual C++") in the past for obscure legal reasons, but now that Clang ships  as part of Visual Studio, clarity is more important than marginal increases in legal risk. This is probably why the code talks about the "Microsoft C++ ABI" not the "MSVC(++) ABI".

AaronBallman replied:
SGTM!

ojhunt replied:
I've switched to Microsoft ABI, and the diagnostic is now `ms-bitfield-padding`

AaronBallman commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -633,6 +633,52 @@ def Packed : DiagGroup<"packed", [PackedNonPod]>;
 def UnalignedAccess : DiagGroup<"unaligned-access">;
+def MSBitfieldCompatibility : DiagGroup<"ms-bitfield-compatibility"> {
+  code Documentation = [{
+    Under the MS Windows platform ABI, adjacent bit-fields are not packed if the
```
```suggestion
    Under the Microsoft ABI, adjacent bit-fields are not packed if the
```

AaronBallman commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -633,6 +633,52 @@ def Packed : DiagGroup<"packed", [PackedNonPod]>;
+        Enum2 field2 : 1;
+      };
+
+    In each of these cases under the MS Windows platform ABI the second bit-field
```
```suggestion
    In each of these cases under the Microsoft ABI the second bit-field
```

ojhunt replied:
Fixed

AaronBallman reviewed (approved):
LGTM with some minor nits about use of MS vs Microsoft.

ojhunt committed:
Merge remote-tracking branch 'origin/llvm.org/main' into users/ojhunt/warn-msvc-bitfield-packing

t push

ojhunt commented:
(Finally getting back to this one)

ojhunt committed:
Change diagnostic spelling, switch to 'Microsoft ABI' to describe the ABI

ojhunt commented:
@AaronBallman are you ok with this now? I'm re-requesting the review as it's been a while, and I want to confirm I addressed things as you wanted

ojhunt committed:
Merge remote-tracking branch 'origin/llvm.org/main' into users/ojhunt/warn-msvc-bitfield-packing
